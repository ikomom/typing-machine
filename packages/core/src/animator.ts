import { calculatePatches, diffString, sliceInput } from './index'
import type { AnimatorStep, Patch } from './index'

export function* animatePatches(input: string, patches: Patch[]): Generator<AnimatorStep> {
  let content = input
  // 当前游标
  let cursor = 0
  // 每次只关注添加、删除的位置, 并且拼接起来
  for (let index = 0; index < patches.length; index++) {
    const patch = patches[index]
    const head = content.slice(0, patch.cursor)

    yield { type: 'new-patch', patch, index }

    if (patch.type === 'insert') {
      // 光标从头部插入
      cursor = patch.cursor
      const tail = content.slice(patch.cursor)
      // 模拟动画过程
      yield { type: 'insert-start', cursor, content }
      for (const { char, output, cursor: delta } of animateInsertionSlices(patch.content)) {
        yield {
          type: 'insert',
          cursor: cursor + delta,
          content: head + output + tail,
          char,
        }
      }
      content = head + patch.content + tail
      yield { type: 'insert-end', cursor, content }
    }
    else if (patch.type === 'removal') {
      // 光标从尾部开始减小
      cursor = patch.cursor + patch.length
      const tail = content.slice(patch.cursor + patch.length)
      const selection = content.slice(patch.cursor, patch.cursor + patch.length)
      yield { type: 'removal-start', cursor, content }

      for (let i = selection.length - 1; i >= 0; i--) {
        yield {
          type: 'removal',
          cursor: cursor - (selection.length - i),
          content: head + selection.slice(0, i) + tail,
        }
      }
      content = head + tail
      yield { type: 'removal-end', cursor, content }
    }
  }

  yield { type: 'animator-finish', content }
}

export function *animateInsertionSlices(input: string) {
  const slices = sliceInput(input)
  let output = ''
  for (const { content, cursor } of slices) {
    const head = output.slice(0, cursor)
    const tail = output.slice(cursor)

    let body = ''
    for (const char of content) {
      body += char
      // console.log('body', { body, char })
      yield {
        char,
        output: head + body + tail,
        cursor: cursor + body.length,
      }
    }
    output = head + content + tail
  }
}

export function animateTo(input: string, output: string) {
  const delta = diffString(input, output)
  const patches = calculatePatches(delta)
  return animatePatches(input, patches)
}

/**
 * TODO: 用法成疑
 * @param input
 * @param patches
 */
export function applyPatches(input: string, patches: Patch[]) {
  // 每次只关注添加、删除的位置, 并且拼接起来
  for (const patch of animatePatches(input, patches)) {
    if (patch.type === 'animator-finish')
      return patch.content
  }
  return input
}
