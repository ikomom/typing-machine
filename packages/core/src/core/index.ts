export * from './patch'
export * from './animator'
export * from './types'
export * from './snaps'
