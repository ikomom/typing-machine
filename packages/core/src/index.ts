export * from './patch'
export * from './animator'
export * from './types'
export * from './snaps'
export * from './utils'
export * from './slicing'
export * from './timming'
