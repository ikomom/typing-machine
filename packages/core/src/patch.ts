import { DiffMatchPatch, DiffOperation } from 'diff-match-patch-typescript'
import type { Diff } from 'diff-match-patch-typescript'
import type { Patch } from './types'

export function diffString(a: string, b: string): Diff[] {
  const diff = new DiffMatchPatch()
  const res = diff.diff_main(a, b)
  diff.diff_cleanupSemantic(res)
  return res
}

export function calculatePatches(diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  // 当前文字的index
  let cursor = 0
  for (const change of diffs) {
    const [operation, content] = change
    const offset = content.length
    if (operation === DiffOperation.DIFF_EQUAL) {
      cursor += offset
    }
    else if (operation === DiffOperation.DIFF_DELETE) {
      patches.push({
        type: 'removal',
        cursor,
        length: offset,
      })
      // 删除时，游标还是一样，所以不用移动，一下代码不用
      // cursor -= offset
    }
    else if (operation === DiffOperation.DIFF_INSERT) {
      patches.push({
        type: 'insert',
        cursor,
        content,
      })
      // 添加时, 游标 = 之前游标 + 添加长度
      cursor += offset
    }
    else {
      throw new Error('unknown change type')
    }
  }
  return patches
}
