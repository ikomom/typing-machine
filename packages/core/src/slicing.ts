import type { Slice } from './types'

const pairs = [
  ['{', '}'],
  ['(', ')'],
  ['[', ']'],
  ['<', '>'],
  ['`', '`'],
  ['\'', '\''],
  ['"', '"'],
]
const chars = '(){}\\[\\]<>\'"`'
const splitRegex = new RegExp(`(\s*[${chars}]\s*)`)

/**
 * slide input by pairs
 * @param input
 */
export function sliceInput(input: string): Slice[] {
  if (!input)
    return []
  // const rest = input.split(/\r?\n/).join('')
  const tailingNewLines = input.match(/\n+$/m)?.[0] || ''
  const rest = tailingNewLines.length
    ? input.slice(0, -tailingNewLines.length)
    : input

  // string slice
  const strings = rest.split(splitRegex).filter(Boolean)

  let index = 0
  const slices = strings
    .map((s, idx) => {
      const trimmed = s.trim()
      const pair = pairs.find(i => i[1] === trimmed)
      // pair before 2 index
      const order = (pair && strings[idx - 2]?.trim() === pair[0]) ? idx - 2 : idx

      const _index = index
      index += s.length

      return {
        content: s,
        order,
        cursor: _index,
      }
    }).sort((a, b) => a.order - b.order)

  let cursor = 0
  slices.forEach((i) => {
    i.cursor = Math.min(i.cursor, cursor)
    cursor += i.content.length
  })
  if (tailingNewLines.length) {
    slices.unshift({
      content: tailingNewLines,
      order: -1,
      cursor: 0,
    })
  }
  // console.log({ slices: [...slices], strings })
  return slices
}
