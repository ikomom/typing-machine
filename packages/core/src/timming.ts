export const timeRanges: [string, number, number][] = [
  // default
  ['', 0, 100],
  // commons
  [' ', 0, 40],
  ['adefijklosu', 0, 50],
  ['bcghmnpqrtvwxyz', 5, 70],
  // uppercase
  ['adefijklosu'.toUpperCase(), 0, 50],
  ['bcghmnpqrtvwxyz'.toUpperCase(), 5, 70],
  // a bit far
  ['[];\',./-=', 10, 120],
  // need shift
  ['<>?:"{}|\\_+()*', 20, 140],
  // a bit far and shift
  ['!~`#$%^&', 20, 180],
]

export function randRange(min: number, max: number): number {
  return Math.random() * (max - min) + min
}

export function getTimeout(char: string, multiplier = 1): number {
  const [, min, max] = timeRanges.find(range => range[0].includes(char)) || timeRanges[0]
  return randRange(min, max) * multiplier
}
