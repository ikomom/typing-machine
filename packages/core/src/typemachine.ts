import { getTimeout, randRange } from './timing'
import type { AnimatorStep } from './types'

export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export interface TypewriterOptions {

}

export async function *typingAnimator(
  steps: Generator<AnimatorStep>,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  options: TypewriterOptions = {},
): AsyncGenerator<AnimatorStep> {
  for (const step of steps) {
    switch (step.type) {
      case 'init':
        break
      case 'new-snap':
        if (step.index)
          await sleep(randRange(700, 1000))
        break
      case 'new-patch':
        if (step.index)
          await sleep(randRange(200, 500))
        break
      case 'insert':
        await sleep(getTimeout(step.char, 1.2))
        break
      case 'removal':
        await sleep(randRange(0, 5))
        break
    }
    yield step
  }
}
