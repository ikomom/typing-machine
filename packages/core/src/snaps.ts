import { simpleAnimator } from './animator'
import type { AnimatorStep, Snapshot } from './types'
import { replaceAll } from './utils'

export const SNAP_HEADING = 'ik-typing-machine Snapshots v1'
export const SNAP_SEPARATOR_PRE = '-'.repeat(2)
export const SNAP_SEPARATOR_SUFFIX = '-'.repeat(10)
export const SNAP_SEPARATOR_OPTIONS = '----OPTIONS----'
export const SNAP_SEPARATOR = `${SNAP_SEPARATOR_PRE}--${SNAP_SEPARATOR_SUFFIX}`
export const snapSeparatorMatcher = (eol: string) => new RegExp(`${eol}?${SNAP_SEPARATOR_PRE}[#\\w-]*${SNAP_SEPARATOR_SUFFIX}${eol}`, 'g')
export const snapSeparatorMatcherOptions = (eol: string) => new RegExp(`${eol}?${SNAP_SEPARATOR_OPTIONS}`, 'g')

/**
 * Snapshot store
 */
export class Snapshots extends Array<Snapshot> {
  constructor(...props: Snapshot[]) {
    super(...props)
  }

  get last() {
    return this[this.length - 1]
  }

  get first() {
    return this[0]
  }

  toString(eol = '\n'): string {
    return [
      SNAP_HEADING + eol,
      ...this.flatMap((snap, i) => {
        return [
          `${SNAP_SEPARATOR_PRE + String(i + 1).padStart(2, '0') + SNAP_SEPARATOR_SUFFIX}`,
          snap.content,
          ...(snap.options
            ? [
                SNAP_SEPARATOR_OPTIONS,
                Object.keys(snap.options).length > 1
                  ? JSON.stringify(snap.options, null, 2)
                  : JSON.stringify(snap.options)]
            : []),
        ]
      }),
      SNAP_SEPARATOR,
      '',
    ].join(eol)
  }

  static fromRawStr(raw: string) {
    const parts = raw.split('\r?\n')
  }

  static fromString(raw: string, eol = '\n') {
    let parts = raw
      .split(snapSeparatorMatcher(eol))
      // .slice(1, -1) // remove header and tailing
    // console.log({ raw, q: replaceAll(raw, eol, '\n'), eol, parts })
    parts = parts.slice(1, -1)
    const snapshots: Snapshot[] = []
    for (let i = 0; i < parts.length; i += 1) {
      const [content, optionsRaw] = parts[i].split(snapSeparatorMatcherOptions(eol))
      const snap: Snapshot = {
        content,
      }
      if (optionsRaw?.trim())
        snap.options = JSON.parse(optionsRaw)

      snapshots.push(snap)
    }

    return new Snapshots(...snapshots)
  }

  *animate(): Generator<AnimatorStep> {
    let lastContent: string | undefined
    const copy: Snapshot[] = [...this]

    for (let index = 0; index < copy.length; index++) {
      const snap = copy[index]
      if (lastContent === undefined) {
        lastContent = snap.content
        yield {
          type: 'init',
          content: lastContent,
        }
        continue
      }

      yield {
        type: 'new-snap',
        snap,
        index,
      }
      // form last to current, generate animator array
      const animator = simpleAnimator(lastContent, snap.content)
      for (const result of animator)
        yield result

      lastContent = snap.content
    }
  }
}

export type SnapshotFallbackLoader = (id: string) => Snapshots | undefined | Promise<Snapshots | undefined>
export interface SnapshotManagerOptions {
  ensureFallback?: SnapshotFallbackLoader
}

export class SnapshotManager extends Map<string, Snapshots> {
  constructor(
    public options: SnapshotManagerOptions,
  ) {
    super()
  }

  /**
   * set snapshot by id, if no present, read load callback or create new snapshot
   * @param id
   * @param load
   */
  async ensure(id: string, load = this.options.ensureFallback): Promise<Snapshots> {
    if (!this.has(id))
      this.set(id, await load?.(id) || new Snapshots())

    return this.get(id)!
  }
}
