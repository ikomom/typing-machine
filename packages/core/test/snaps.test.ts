import { describe, expect, it } from 'vitest'
import type { Snapshot } from '../src'
import { Snapshots } from '../src'

describe('snaps', () => {
  it('persist', () => {
    const data: Snapshot[] = [
      { content: '', options: { wait: 200 } },
      { content: 'import {} from "vue"', options: { wait: 100 } },
      { content: 'import { createApp } from "vue"\n\nconst app = createApp()\n', options: { wait: 100 } },
    ]

    const snaps = new Snapshots(...data)
    const serialized = snaps.toString()

    expect(serialized).toMatchSnapshot('serialized')

    const deserialized = Snapshots.fromString(serialized)

    expect(data).toEqual([...deserialized])
  })
})
