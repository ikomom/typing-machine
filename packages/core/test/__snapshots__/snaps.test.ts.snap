// Vitest Snapshot v1

exports[`snaps > deserialized > input 1`] = `
"ik-typing-machine Snapshots v1

--01----------

--02----------

--03----------
import App from './App.vue'

--04----------
import { } from 'vue'
import App from './App.vue'
----OPTIONS----
{\\"wait\\":200,\\"pause\\":false}
--05----------
import { createApp } from 'vue'
import App from './App.vue'
const app = createApp(App)
app.mount('#app')

--06----------
import { createApp } from 'vue'
import { createRouter } from 'vue-router'
import App from './App.vue'
const app = createApp(App)
const router = createRouter()
app.use(router)
app.mount('#app')

--07----------
import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import App from './App.vue'
const app = createApp(App)
const router = createRouter({
  history: createWebHistory(),
})
app.use(router)
app.mount('#app')
----OPTIONS----
{\\"wait\\":1200,\\"pause\\":true}
--08----------
import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import routes from 'virtual:generated-pages'
import App from './App.vue'
const app = createApp(App)
const router = createRouter({
  history: createWebHistory(),
  routes,
})
app.use(router)
app.mount('#app')

--------------
"
`;

exports[`snaps > deserialized > output 1`] = `
"ik-typing-machine Snapshots v1

--01----------

--02----------

--03----------
import App from './App.vue'

--04----------
import { } from 'vue'
import App from './App.vue'
----OPTIONS----
{\\"wait\\":200,\\"pause\\":false}
--05----------
import { createApp } from 'vue'
import App from './App.vue'
const app = createApp(App)
app.mount('#app')

--06----------
import { createApp } from 'vue'
import { createRouter } from 'vue-router'
import App from './App.vue'
const app = createApp(App)
const router = createRouter()
app.use(router)
app.mount('#app')

--07----------
import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import App from './App.vue'
const app = createApp(App)
const router = createRouter({
  history: createWebHistory(),
})
app.use(router)
app.mount('#app')
----OPTIONS----
{\\"wait\\":1200,\\"pause\\":true}
--08----------
import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import routes from 'virtual:generated-pages'
import App from './App.vue'
const app = createApp(App)
const router = createRouter({
  history: createWebHistory(),
  routes,
})
app.use(router)
app.mount('#app')

--------------
"
`;

exports[`snaps > persist > serialized 1`] = `
"ik-typing-machine Snapshots v1

--01----------

----OPTIONS----
{\\"wait\\":200,\\"pause\\":false}
--02----------
import {} from \\"vue\\"
--03----------
import { createApp } from \\"vue\\"

const app = createApp()

----OPTIONS----
{\\"wait\\":100}
--04----------
import { useMemo } from \\"react\\"

const memo = useMemo()

----OPTIONS----
{\\"wait\\":100}
--------------
"
`;
